File: 2058. Find the Minimum and Maximum Number of Nodes Between Critical Points/2058. Find the Minimum and Maximum Number of Nodes Between Critical Points.py
from math import inf
class Solution:        
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        if not head or not head.next:
            return [0, 0]
        
        arr = []
        min_d, max_d = float('inf'), -float('inf')
        first_critical, last_critical = None, None
        while head:
            arr.append(head.val)
            head = head.next
        for i in range(1, len(arr) - 1):
            if (arr[i] > arr[i-1] and arr[i] > arr[i+1]) or (arr[i] < arr[i-1] and arr[i] < arr[i+1]):
                if last_critical is None:
                    first_critical = last_critical = i
                else:
                    min_d = min(min_d, i - last_critical)
                    max_d = max(max_d, i - first_critical)
                    last_critical = i
        if min_d == float('inf') or max_d == -float('inf'):
            return [-1, -1]
        return [min_d, max_d]

File: 2058. Find the Minimum and Maximum Number of Nodes Between Critical Points/2058. Find the Minimum and Maximum Number of Nodes Between Critical Points.cc
class Solution
{
public:
    vector<int> nodesBetweenCriticalPoints(ListNode *head)
    {
        vector<int> critical;
        ListNode *cur = head;
        ListNode *next = head->next;
        ListNode *prev = NULL;
        int index = 0;
        while (next != NULL)
        {

            if ((prev != NULL && cur->val < prev->val && cur->val < next->val) ||
                (prev != NULL && cur->val > prev->val && cur->val > next->val))
            {
                critical.push_back(index);
            }
            prev = cur;
            cur = next;
            next = next->next;
            index++;
        }

        if (critical.size() < 2)
            return {-1, -1};
        int maxDist = critical.back() - critical.front();
        int minDist = INT_MAX;

        for (int i = 0; i < critical.size() - 1; i++)
        {
            minDist = min(minDist, critical[i + 1] - critical[i]);
        }
        return {minDist, maxDist};
    }
};